backbone/src/test/java/org/researchstack/backbone/ui/task/TaskViewModelTest.kt

```diff
@@ -0,0 +1,419 @@
+    }
+
+    @Test
+    fun testNext_shouldCloseWhenNextNull() {
+        // Assemble
+        taskViewModel = spy(createViewModel())
+
+        // Act
+        taskViewModel.nextStep()
+
+        // Assert
+        Assert.assertEquals(taskViewModel.taskCompleted.value, true)
+    }
+
+    @Test
+    fun testShowCurrentStep_ifCurrentStepIsNull_ShowGoToNextStep() {
+        // Assemble
+        taskViewModel = spy(createViewModel())
+        // Act
+        taskViewModel.showCurrentStep()
+
+        // Assert
+        verify(taskViewModel, times(1)).showCurrentStep()
+    }
+
+    @Test
+    fun testPreviousOnEditMode_shouldCheckForCancelDialog() {
+        // Assemble
+        taskViewModel = spy(createViewModel())
+        val currentStepMocked = createCurrentStep()
+        taskViewModel.edit(currentStepMocked)
+
+        // Act
+        taskViewModel.previousStep()
+
+        // Assert
+        verify(taskViewModel, times(1)).showCancelEditAlert()
+    }
+
+    @Test
+    fun testNextStepEditMode_WhenNextStepIsReviewStep_EditShouldBeFalse_moveReviewStepShouldGoToReviewStep() {
+        // Assemble
+        taskViewModel = spy(createViewModel())
+
+        val currentStepMocked = createCurrentStep()
+        val nextStepMocked = createCurrentStep("nextStepMocked")
+        taskViewModel.currentStep = currentStepMocked
+
+        `when`(task!!.getStepAfterStep(any(), any())).thenReturn(nextStepMocked)
+        `when`(taskViewModel.isReviewStep(nextStepMocked)).thenReturn(true)
+        taskViewModel.edit(nextStepMocked)
+
+        // Act
+        taskViewModel.nextStep()
+
+        // Assert
+        Assert.assertEquals(taskViewModel.editing, false)
+        Assert.assertEquals(taskViewModel.moveReviewStep.value?.step, nextStepMocked)
+    }
+
+    @Test
+    fun testCheckIfAnswersAreNotTheSame_false() {
+        // Assemble
+        taskViewModel = spy(createViewModel())
+        val clonedTaskResultInCaseOfCancel = Mockito.mock(TaskResult::class.java)
+        val currentTaskResult = Mockito.mock(TaskResult::class.java)
+
+        val nextStepMocked = createCurrentStep("nextStepMocked")
+        taskViewModel.currentStep = nextStepMocked
+
+        val originalStepResult = spy(StepResult<String>(nextStepMocked))
+        originalStepResult.result = "tow"
+
+        `when`(clonedTaskResultInCaseOfCancel!!.getStepResult(any())).thenReturn(originalStepResult)
+        `when`(currentTaskResult!!.getStepResult(any())).thenReturn(originalStepResult)
+        `when`(taskViewModel.currentTaskResult).thenReturn(currentTaskResult)
+        taskViewModel.clonedTaskResultInCaseOfCancel = clonedTaskResultInCaseOfCancel
+
+        // Act
+        val actualResult = taskViewModel.checkIfAnswersAreTheSame()
+
+        // Assert
+        Assert.assertEquals(actualResult, false)
+    }
+
+    @Test
+    fun testCheckIfAnswersAreNotTheSame_true() {
+        // Assemble
+        taskViewModel = spy(createViewModel())
```

Comment:
same as above, and below in the testCheckIfNewAnserIsSkip... etc.etc.etc. All these `spy` on the subject are _suspicious_ :)